python/herbstluftwm/types.py

```diff
@@ -48,3 +48,81 @@ def adjusted(self, dx=0, dy=0, dw=0, dh=0):
         are adjusted by the provided deltas.
         """
         return Rectangle(self.x + dx, self.y + dy, self.width + dw, self.height + dh)
+
+
+class HlwmType:
+    """
+    Wrapper functions for converting between python types and types
+    in herbstluftwm.
+    """
+    def __init__(self, name, from_user_str, to_user_str, is_instance):
+        # a hlwm type should define the following
+        self.name = name  # type: str
+        # a callback for parsing
+        self.from_user_str = from_user_str  # of type: str -> T
+        # a callback for printing
+        self.to_user_str = to_user_str  # of type: T -> str
+        # a predicate whether a python variable has this type:
+        self.is_instance = is_instance  # of type: Anything -> bool
+
+    @staticmethod
+    def by_name(type_name):
+        """Given the full name of a hlwm type, return
+        the metadata
+        python type"""
+        for t in hlwm_types():
+            if t.name == type_name:
+                return t
+        return None
+
+    @staticmethod
+    def by_type_of_variable(python_variable):
+        """Given a variable, detect its type
+        """
+        for t in hlwm_types():
+            if t.is_instance(python_variable):
+                return t
+        return None
+
+
+def hlwm_types():
+    """
+    Return a list of HlwmType objects.
+
+    Unfortunately, the order matters for the is_instance() predicate: Here, the
+    first matching type in the list must be used. (This is because
+    `isinstance(True, int)` is true)
+    """
+    types = [
+        HlwmType('bool',
+                 bool_from_user_str,
+                 lambda b: 'true' if b else 'false',
+                 lambda x: isinstance(x, bool)),
+
+        HlwmType('int', int, str, lambda x: isinstance(x, int)),
+
+        # there is no uint in python, so we just convert it to 'int'
+        HlwmType('uint', int, str, lambda x: False),
+
+        HlwmType('rectangle',
+                 Rectangle.from_user_str,
+                 Rectangle.to_user_str,
+                 lambda x: isinstance(x, Rectangle)),
+
+        HlwmType('string',
+                 lambda x: x,
+                 lambda x: x,
+                 lambda x: isinstance(x, str)),
```

Comment:
Personally I find this a bit hard to follow, because the positional lambdas are different "methods", and then there's things scattered in different places in this file (e.g. the `bool_from_user_str` below).

Have you considered making those subclasses of `HlwmType` (with the name being a constant class attribute, and the lambdas being methods instead)?

Reply:
I considered this, but it felt odd to have subclasses with `@staticmethod`s only (and undefined `@staticmethod`s in `HlwmType`). Also I had the impression that it's more write effort.

I agree  that the positional arguments are hard to read. I have just replaced them with named arguments; is this more readable?

Reply:
Definitely better! I'd still prefer classes (then you could also e.g. have proper type annotations instead of describing the types in the comments), but no strong opinions either way.